src/v/cluster/controller_backend.cc

```diff
@@ -635,21 +635,27 @@ controller_backend::calculate_learner_initial_offset(
 
     const auto last_uploaded_to_cloud
       = p->archival_meta_stm()->manifest().get_last_offset();
-
+    const auto manifest_clean_offset
+      = p->archival_meta_stm()->get_last_clean_at();
     /**
      * Last offset uploaded to the cloud is target learner retention upper
      * bound. We can not start retention recover from the point which is not yet
      * uploaded to Cloud Storage.
      */
     vlog(
       clusterlog.trace,
-      "[{}] calculated retention offset: {}, last uploaded to cloud: {}",
+      "[{}] calculated retention offset: {}, last uploaded to cloud: {}, "
+      "manifest clean offset: {}",
       p->ntp(),
       *retention_offset,
-      last_uploaded_to_cloud);
+      last_uploaded_to_cloud,
+      manifest_clean_offset);
+
+    auto const cloud_storage_safe_offset = std::min(
+      last_uploaded_to_cloud, manifest_clean_offset);
```

Comment:
This makes sense, but it's duplicating the logic of `archival_metadata_stm::max_collectible_offset`. You could just call that method instead since it also accounts for the cases when remote write is disabled or the topic is a read replica.

Reply:
done